src/client.rs

```diff
@@ -280,3 +270,368 @@ where
+                    return Err(TransactionError::Dropped.into())
+                }
+                // Ignore and wait for next status event:
+                _ => continue,
+            }
+        }
+        Err(RpcError::Custom("RPC subscription dropped".into()).into())
+    }
+
+    /// Wait for the transaction to be finalized, and for the transaction events to indicate
+    /// that the transaction was successful. Returns the events associated with the transaction,
+    /// as well as a couple of other details (block hash and extrinsic hash).
+    ///
+    /// **Note:** consumes self. If you'd like to perform multiple actions as progress is made,
+    /// use [`TransactionProgress::next()`] instead.
+    pub async fn wait_for_finalized_success(self) -> Result<TransactionEvents<T>, Error> {
+        let evs = self.wait_for_finalized().await?.wait_for_success().await?;
+        Ok(evs)
+    }
+}
+
+/// Possible transaction progess states returned from our [`TransactionProgress::next()`] call.
+#[derive(Debug)]
+pub enum TransactionProgressStatus<'client, T: Config> {
+    /// Transaction is part of the future queue.
+    Future,
+    /// Transaction is part of the ready queue.
+    Ready,
+    /// The transaction has been broadcast to the given peers.
+    Broadcast(Vec<String>),
+    /// Transaction has been included in block with given hash.
+    InBlock(TransactionInBlock<'client, T>),
+    /// The block this transaction was included in has been retracted.
+    Retracted(T::Hash),
+    /// Maximum number of finality watchers has been reached,
+    /// old watchers are being removed.
+    FinalityTimeout(T::Hash),
+    /// Transaction has been finalized by a finality-gadget, e.g GRANDPA
+    Finalized(TransactionInBlock<'client, T>),
+    /// Transaction has been replaced in the pool, by another transaction
+    /// that provides the same tags. (e.g. same (sender, nonce)).
+    Usurped(T::Hash),
+    /// Transaction has been dropped from the pool because of the limit.
+    Dropped,
+    /// Transaction is no longer valid in the current state.
+    Invalid,
+}
+
+impl<'client, T: Config> TransactionProgressStatus<'client, T> {
+    /// A convenience method to return the `Finalized` details. Returns
+    /// [`None`] if the enum variant is not [`TransactionProgressStatus::Finalized`].
+    pub fn as_finalized(&self) -> Option<&TransactionInBlock<'client, T>> {
+        match self {
+            Self::Finalized(val) => Some(val),
+            _ => None,
+        }
+    }
+
+    /// A convenience method to return the `InBlock` details. Returns
+    /// [`None`] if the enum variant is not [`TransactionProgressStatus::InBlock`].
+    pub fn as_in_block(&self) -> Option<&TransactionInBlock<'client, T>> {
+        match self {
+            Self::InBlock(val) => Some(val),
+            _ => None,
+        }
+    }
+}
+
+/// This struct represents a transaction that has made it into a block.
+pub struct TransactionInBlock<'client, T: Config> {
```

Comment:
Would it make sense to provide an easy way for users to query the finalization status of one of these? Imagine a scenario where the users' app takes an optimistic stance, awaits "InBlock" status and then prepares things on its end anticipating finalization and then, before enacting whatever the app does when the extrinsic was successful, it checks if the block is actually final. Maybe `is_final()`?

Reply:
At some point I looked for such a method, but offhand I haven't found a way to ask a node whether a specific block has been finalized (without getting the finalized head and iterating back!).

The best way to handle this use case is to use the "lower level" API exposed here and do something like:

```
while let Some(status) = transaction_progress.next().await? {
    if let Some(finalized) = status.as_in_block() {
        // made it into a block.. prepare some things
    }
    if let Some(finalized) = status.as_finalized() {
        // it's been finalized. we can check block hash is same as above and complete.
    }
}
```